python/lsst/pipe/base/pipeline.py

```diff
@@ -223,52 +222,22 @@ def subsetFromLabels(self, labelSpecifier: LabelSpecifier) -> Pipeline:
         than it should.
         """
-
-        pipeline = copy.deepcopy(self)
-
-        def remove_contracts(label: str):
-            """Remove any contracts that contain the given label
-
-            String comparison used in this way is not the most elegant and may
-            have issues, but it is the only feasible way when users can specify
-            contracts with generic strings.
-            """
-            new_contracts = []
-            for contract in pipeline._pipelineIR.contracts:
-                # match a label that is not preceded by an ASCII identifier, or
-                # is the start of a line and is followed by a dot
-                if re.match(f".*([^A-Za-z0-9_]|^){label}[.]", contract.contract):
-                    continue
-                new_contracts.append(contract)
-            pipeline._pipelineIR.contracts = new_contracts
-
-        # Labels supplied as a set, explicitly remove any that are not in
-        # That list
+        # Labels supplied as a set
         if labelSpecifier.labels:
-            # verify all the labels are in the pipeline
-            if not labelSpecifier.labels.issubset(pipeline._pipelineIR.tasks.keys()):
-                difference = labelSpecifier.labels.difference(pipeline._pipelineIR.tasks.keys())
-                raise ValueError("Not all supplied labels are in the pipeline definition, extra labels:"
-                                 f"{difference}")
-            # copy needed so as to not modify while iterating
-            pipeline_labels = list(pipeline._pipelineIR.tasks.keys())
-            for label in pipeline_labels:
-                if label not in labelSpecifier.labels:
-                    pipeline.removeTask(label)
-                    remove_contracts(label)
+            labelSet = labelSpecifier.labels
         # Labels supplied as a range, first create a list of all the labels
         # in the pipeline sorted according to task dependency. Then only
         # keep labels that lie between the supplied bounds
         else:
-            # Use a dict for fast searching while preserving order
-            # save contracts and remove them so they do not fail in the
-            # expansion step, will be restored after. This is needed because
-            # a user may only configure the tasks they intend to run, which
-            # may cause some contracts to fail if they will later be dropped
-            contractSave = pipeline._pipelineIR.contracts
+            # Create a copy of the pipeline to use when assessing the label
+            # ordering. Use a dict for fast searching while preserving order.
+            # Remove contracts so they do not fail in the expansion step. This
+            # is needed because a user may only configure the tasks they intend
+            # to run, which may cause some contracts to fail if they will later
+            # be dropped
+            pipeline = copy.copy(self)
```

Comment:
Are you sure a one-level deep shallow copy is the right level of copying here?

(I always find `copy.copy` to be scary, and I might go so far as to call it an antipattern, because whether it is or is not the right amount of copying to do is really an implementation detail in general.)

Reply:
Yeah, I had deep copy then switched to copy because I knew the implementation details (there are only a few top level containers of things that nothing really modifies) but it probably is better to always deep copy to avoid any future issues that may arise, it certainly is worth the small amount of memory overhead.